dev/UndockedRegFreeWinRT/catalog.cpp

```diff
@@ -0,0 +1,427 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+#include <pch.h>
+
+#include "catalog.h"
+#include "TypeResolution.h"
+
+#include <activation.h>
+#include <shlwapi.h>
+
+#include <wrl.h>
+
+#include <../DynamicDependency/MddWinRT.h>
+
+using namespace std;
+using namespace Microsoft::WRL;
+
+//TODO Components won't respect COM lifetime. workaround to get them in the COM list? See dev\UndockedRegFreeWinRT\catalog.cpp
+
+extern "C"
+{
+    typedef HRESULT(__stdcall* activation_factory_type)(HSTRING, IActivationFactory**);
+}
+
+// Intentionally no class factory cache here. That would be excessive since
+// other layers already cache.
+struct component
+{
+    wstring module_name;
+    wstring xmlns;
+    HMODULE handle = nullptr;
+    activation_factory_type get_activation_factory;
+    ABI::Windows::Foundation::ThreadingType threading_model;
+
+    ~component()
+    {
+        if (handle)
+        {
+            FreeLibrary(handle);
+        }
+    }
+
+    HRESULT LoadModule()
+    {
+        if (handle == nullptr)
```

Comment:
what is the threading model here?  Can multiple threads travel through here simultaneously?

Reply:
Threading model - URFW is lock-free

Concurrency here - Yes

There's nothing to prevent 2 threads concurrently calling GetActivationFactory() or ActivateInstance() with the same ActivatableClassId. If they're the first for that ACID and you get lucky, both will call here. Both calling LoadLibrary will get the same handle, the 2nd overwriting this->handle written by the first but the handle will be the same. Problem is the destructor will fire once clearing the handle resulting in 2x LoadLibrary but 1x FreeLibrary. That's a best case, could be weirder depending on memory writes and whatnot. There's also only add logic, no removal/unload (as comments like line 19 imply). An obscure case but non-zero.

This is unforked URFW support of WinRT via SxS.

PackageGraph support's more complex so I couldn't merely use this, thus WinRTModuleManager with a thread-safe design (modulo the open issues being followed up on).

Whether or not removal scenarios are addressed for SxS, they need to be for WinRT types via PackageGraph. I need to talk to @kythant and @BenJKuhn before assessing how.